src/nanotron/models/llama.py

```diff
@@ -757,6 +763,125 @@ def forward(
         }
 
 
+class DominoLlamaDecoderLayer(_BaseLlamaDecoderLayer):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        assert self.stream_manager is not None, "DominoLlamaDecoderLayer requires a stream_manager"
+
+    def _core_forward(
+        self,
+        hidden_states: Union[torch.Tensor, TensorPointer],
+        sequence_mask: Union[torch.Tensor, TensorPointer],
+    ) -> List[Union[torch.Tensor, TensorPointer]]:
+        num_input_batches = self.parallel_config.domino.num_input_batches
+
+        orig_sequence_mask = sequence_mask
+        comm_stream = self.stream_manager.get_default_comm_stream()
+        comm_bucket = self.stream_manager.comm_bucket
+
+        hidden_states = torch.chunk(hidden_states, chunks=num_input_batches, dim=1)
+        sequence_mask = torch.chunk(sequence_mask, chunks=num_input_batches, dim=0)
+
+        hidden_states0, hidden_states1 = hidden_states
+        sequence_mask0, sequence_mask1 = sequence_mask
+
+        residual0 = hidden_states0
+        residual1 = hidden_states1
+
+        hidden_states0 = self.input_layernorm(hidden_states0)
+        hidden_states1 = self.input_layernorm(hidden_states1)
+        hidden_states0 = insert_backward_sync_to_tensor(
+            hidden_states0,
+            BWD_ATTN_OP_NAME.format(self.layer_idx, 1),
+            self.stream_manager,
+        )
+        hidden_states1 = insert_backward_sync_to_tensor(
+            hidden_states1,
+            BWD_MLP_OP_NAME.format(self.layer_idx, 0),
+            self.stream_manager,
+        )
+
+        attn_output0 = self.attn(
+            hidden_states=hidden_states0,
+            sequence_mask=sequence_mask0,
+            op_name=FWD_ATTN_OP_NAME.format(self.layer_idx, 0),
+        )
+        attn_output1 = self.attn(
+            hidden_states=hidden_states1,
+            sequence_mask=sequence_mask1,
+            op_name=FWD_ATTN_OP_NAME.format(self.layer_idx, 1),
+        )
+
```

Comment:
would be veeery nice to have context manager instead of passing optional args in the forward 🙏🏼

def _core_forward(self, hidden_states, sequence_mask):
    # ...
    with OpNameContext(FWD_ATTN_OP_NAME.format(self.layer_idx, 0)):
        attn_output0 = self.attn(hidden_states=hidden_states0, sequence_mask=sequence_mask0)
    
    with OpNameContext(FWD_ATTN_OP_NAME.format(self.layer_idx, 1)):
        attn_output1 = self.attn(hidden_states=hidden_states1, sequence_mask=sequence_mask1)
    # ...